keps/sig-release/20190118-creating-packages.md

```diff
@@ -0,0 +1,417 @@
+- *Risk*: `nfpm` is not the right tool, it goes away, works differently, ...  
+  *Mitigations*:  
+  - we encapsulate the build of exactly one package, if things change, we can
+    change the internal implementation but keep the external
+    interface
+- *Risk*: We cannot build proper windows packages but community wants/needs them  
+  *Mitigations*:
+  - Talk to people in sig windows, esp. about packaging their components
+
+## Design Details
+
+### Test Plan
+
+- different components will have unit tests as much as possible
+- comparing output debs/rpms to what currently is released
+  - not sure how? by the files they deploy?
+
+### Graduation Criteria
+
+#### From `nil` to `k/k`, Introduce the new mechanism into `k/k`
+
+- Iterate here on the KEP
+- Do PoCs
+- Compare to packages generated with the current release tooling
+- See no significant difference in the generated packages
+- The new mechanism has means to prepare the packages in a way that they can be
+  published via the current release workflow
+
+#### Deprecate all other options to build packages
+
+Deprecation of other packaging mechanisms is done either by a note in the
+README or even in the source code to print a warning when building packages via
+deprecated tools.
+
+Before deprecation the following points should be implemented:
+
+- Make it easy for the release team to switch between the old and the new process
+- Make it easy for the release team to compare two (sets of) packages and
+  assess if they are sufficiently similar
+- Change the release (branch management & patch management) processes to use
+  the new mechanism by default
+
+#### Make this the only available way to build packages
+
+- X (TBD) releases have been cut with the new machanism
+- There was no feedback on bad packages
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+
+N/A
+
+<!--
+**Note:** *Section not required until targeted at a release.*
+
+If applicable, how will the component be upgraded and downgraded?  Make sure this is in the test plan.
+-->
+
+### Version Skew Strategy
+
+One of the issues we saw is that `k/k` has release branches, while `k/release`
+didn't have the same branch/release strategy.
+This is fixed by adding the new tooling and package building specs to `k/k`
+directly, it will be versioned alongside the things it will package.
+
+## Implementation History
+
+<!--
+Major milestones in the life cycle of a KEP should be tracked in `Implementation History`.
+Major milestones might include
+
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks [optional]
+
+<!-- Why should this KEP _not_ be implemented. -->
```

Comment:
Playing devil's advocate: Have we considered if debs / rpms are something the project should continue to officially publish? What about other package formats? vm images?

What if we just shipped self-contained binaries? The packages today are all just a binary + a systemd configuration AFAIK.

Reply:
> Have we considered if debs / rpms are something the project should continue to officially publish

💯 
I saw that people brought this up but it never really got discussed much as far as I can see. But this is probably THE question to answer. I will make this and some ideas how to validate that a thing in the KEP!

> What about other package formats? vm images?

Assuming we have a good enough abstraction with the thing I called `package configuration` I think that would be doable. Discovering that people actually want `vmdk`s rather then `deb`s is harder to validate.

> What if we just shipped self-contained binaries? The packages today are all just a binary + a systemd configuration AFAIK.

 Thinks I can think of which a package brings (in addition to a single binary):
- the single binary
- the systemd config
- some sysctl configs (?)
- the dependency configuration (which is a whole other topic), e.g.: `kubeadm v1.6` depends on `kubelet >= v1.5 < v1.7`)

Reply:
> Have we considered if debs / rpms are something the project should continue to officially publish

>  Discovering that people actually want vmdks rather then debs is harder to validate.

in SIG cluster lifecycle meetings we brought this topic multiple times.

debs and rpms are a burden that project creators hardly want to deal with.
unfortunately this is something the end user of Linux is used to, expects and we have to support.

so even if we go for things like Snap, people would still want deps and rpms.
it's a Linux ecosystem problem.

Reply:
Yes but usually unless it's a commercial project, especially large projects usually have packages maintained by the linux distro, right?